[Android] Add NetworkChange implementation using PeriodicTimer

simonrozsival opened the pull request:
We can't use netlink sockets to monitor network changes on Android anymore due to changes to permissions in recent Android versions. There doesn't seem to be a native Android API that could be used to implement the `NetworkChange` API in .NET. Instead, this PR implements the functionality using `PeriodicTimer` and manually diffing the IP addresses of all network interfaces. The period is set to 2s to limit the impact on battery life of mobile devices.

Ref #77822

simonrozsival committed:
Add NetworkChange implementation using PeriodicTimer

simonrozsival committed:
Code cleanup

simonrozsival commented:
/azp run runtime-android

steveisok reviewed (approved):
LGTM. I think a 2 sec poll is reasonable for those who want to opt in. I don't think there are any other good alternatives.

mdh1418 commented on src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs:
```diff
@@ -0,0 +1,251 @@
+                {
+                    lock (s_lockObj)
+                    {
+                        bool hadSubscribers = s_addressChangedSubscribers.Count != 0 ||
```
How come we need to check if there are address change subscribers here when only s_availabilityChangedSubscribers.Count can decrement in this scope? (In `NetworkAddressChanged` remove, it doesnt check for availability change subscribers)

simonrozsival replied:
I took this code as is from the `NetworkAddressChange.OSX.cs` implementation. I considered it "battle tested" and so I didn't make any changes.

It would be better if we shared the implementation in the common partial class file, but I didn't want to touch the existing code in this PR if possible.

mdh1418 replied:
Sounds good, I didnt think it really changes the logic, as ultimately the boolean seems to ensure theres only one call to `StopLoop` in case there were successive calls. I didn't realize it was in `NetworkAddressChange.OSX.cs`, and I was confused why the remove logic were slightly different :)

mdh1418 commented on src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs:
```diff
@@ -0,0 +1,251 @@
+            s_lastIpAddresses = null;
+        }
+
+        private static async Task PeriodicallyCheckIfAddressChanged(CancellationToken cancellationToken)
```
Since this checks for both NetworkAddress and NetworkAvailability changes, should there be renaming to `PeriodicallyCheckIfNetworkChanged`, `CheckIfNetworkChanged`, `OnNetworkChanged`, etc?

simonrozsival replied:
Thanks for those suggestions. You're right that it'll make more sense if those functions names are more general.

mdh1418 reviewed (approved):
A few nits, but overall looks good to me!
Out of curiosity, were there statistics to use 2 seconds polling? I'm not familiar, but is there a general frequency to network changes, and is the impact on battery life mainly from `CheckIfAddressChanged` firing? How long does it take for `CheckIfAddressChanged` to finish?

simonrozsival commented on src/libraries/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj:
```diff
@@ -176,6 +176,9 @@
     <Compile Include="$(CommonPath)Interop\BSD\System.Native\Interop.ProtocolStatistics.cs" Link="Common\Interop\BSD\System.Native\Interop.ProtocolStatistics.cs" />
     <Compile Include="$(CommonPath)Interop\BSD\System.Native\Interop.TcpConnectionInfo.cs" Link="Common\Interop\BSD\System.Native\Interop.TcpConnectionInfo.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'android'">
```
There are separate ItemGroups for network interfaces on different platforms and then there are ItemGroups for NetworkAddressChange and I followed the existing pattern. I don't have a preference and I think it's OK both ways. I'll merge those two Android item groups 👍.

simonrozsival committed:
Rename functions

simonrozsival committed:
Update csproj

simonrozsival commented:
@mdh1418 The 2s is my estimate of what is "almost immediately" but at the same time "not all the time". I tried being as conservative as possible with the frequency.

I didn't measure how long ~`CheckIfAddressChanged`~ `CheckIfNetworkChanged` runs because it will be different on different devices. The function enumerates the network interfaces present on the phone/tablet and that will behave a little differently on each device. I can make some measurements tomorrow.

mdh1418 commented:
Gotcha, I was wondering if the impact of battery life would be more directly related to the duty cycle of `CheckIfNetworkChanged`. If it takes a whole second to run (50% duty cycle/active time), then 2 seconds or even longer seems to make sense (without considering how important it is to not miss network changes), and on the other hand if it takes 1ns to run, then we can probably check more frequently. But I'm not sure if thats exactly how the battery is directly impacted.

What might be more helpful to determine polling frequency than duty cycle might be what happens if we miss a network change? Is it problematic if there were multiple network changes within the 2 seconds? Is it problematic if there was a change that cancelled itself out within 2 seconds (i.e. a network becomes available and then becomes unavailable all before the next poll)?

simonrozsival commented:
@mdh1418 So on my Samsung Galaxy phone running Android 13 `CheckIfNetworkChanged` usually takes around 2-8 milliseconds when there isn't any change since the previous call and around 40-50 ms whenever there is a change. So it isn't on the order of seconds, but neither is it on the order of nanoseconds.

I don't think it's too important to detect these short network availability spikes, although there might be customers who might need this kind of information. The use case for this API in a mobile app I can imagine is to wait for network availability and only when the device is online, attempt to send a HTTP request. This implementation would IMO be suitable for that use case. it wouldn't be useful to monitor and diagnose network stability of an Android device. I think there are more suitable native APIs for that use case.

simonrozsival commented:
/azp run runtime-android

simonrozsival commented:
Failing tests are unrelated to this PR.

simonrozsival commented:
/backport to release/7.0
